### PR TITLE
Add TempoCNN, auxiliary algos and unit tests

### PR DESCRIPTION
**Pull request by pabloEntropia**
- TensorflowInputTempoCNN → Produce mel bands similar to
https://github.com/hendriks73/tempo-cnn/blob/master/tempocnn/feature.py

  - TensorflowPredictTempoCNN → Make predictions with TempoCNN models

  - TempoCNN → Postprocess predictions to get a global tempo estimation

### CODE REVIEW

File: src/algorithms/rhythm/tempocnn.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+#include "tempocnn.h"
+
+using namespace std;
+using namespace essentia;
+using namespace standard;
+
+const char* TempoCNN::name = "TempoCNN";
+const char* TempoCNN::category = "Rhythm";
+const char* TempoCNN::description = DOC(
+  "This algorithm makes predictions using Hendrik Schreiber's TempoCNN-based models.\n"
```
Comment (dbogdanov) [inline src/algorithms/rhythm/tempocnn.cpp:29]: Same here.

Comment (dbogdanov) [inline src/algorithms/rhythm/tempocnn.cpp:29]: The first sentence (short description) of TempoCNN is equivalent to TensorflowPredictTempoCNN. They should be distinguished somehow. 

TensorflowPredictTempoCNN predicts from a given input mel-spectrogram, while TempoCNN predicts from a given input audio.

File: src/algorithms/rhythm/tempocnn.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  "Note: This algorithm does not make any check on the input model so it is the user's responsibility to make sure it is a valid one."
+  "\n"
+  "References:\n"
+  "  [1] Hendrik Schreiber, Meinard Müller, A Single-Step Approach to Musical Tempo Estimation Using a Convolutional Neural Network Proceedings of the 19th International Society for Music Information Retrieval Conference (ISMIR), Paris, France, Sept. 2018.\n"
+  "  [2] Hendrik Schreiber, Meinard Müller, Musical Tempo and Key Estimation using Convolutional Neural Networks with Directional Filters Proceedings of the Sound and Music Computing Conference (SMC), Málaga, Spain, 2019.\n"
+  "  [3] Original models and code at https://github.com/hendriks73/tempo-cnn\n"
+  "  [4] Supported models at https://essentia.upf.edu/models/");
+
+
+// Hendrik's models output a softmax distribution of length 256.
```
Comment (dbogdanov) [inline src/algorithms/rhythm/tempocnn.cpp:46]: May be we should refer to TempoCNN in the comments or just "the models", especially if he creates more models in the future.

File: src/algorithms/rhythm/tempocnn.h [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  ~TempoCNN() {
+    if (_tensorflowPredictTempoCNN) delete _tensorflowPredictTempoCNN;
+  }
+
+  void declareParameters() {
+    declareParameter("graphFilename", "the name of the file containing the model to use", "", Parameter::STRING);
+    declareParameter("input", "the name of the input node in the TensorFlow graph", "", "input");
+    declareParameter("output", "the name of the node from which to retrieve the output tensors", "", "output");
+    declareParameter("patchHopSize", "the number of frames between the beginnings of adjacent patches. 0 to avoid overlap", "[0,inf)", 128);
+    declareParameter("lastPatchMode", "what to do with the last frames. Options are to `repeat` them to fill the last patch or to `discard` them", "{discard,repeat}", "discard");
```
Comment (dbogdanov) [inline src/algorithms/rhythm/tempocnn.h:62]: Same here for repeat vs discard.

File: src/algorithms/machinelearning/tensorflowpredicttempocnn.cpp [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+const char* TensorflowPredictTempoCNN::description = essentia::standard::TensorflowPredictTempoCNN::description;
+
+
+TensorflowPredictTempoCNN::TensorflowPredictTempoCNN() : AlgorithmComposite(),
+    _frameCutter(0), _tensorflowInputTempoCNN(0), _vectorRealToTensor(0), _tensorNormalize(0),
+    _tensorTranspose(0), _tensorToPool(0), _tensorflowPredict(0), _poolToTensor(0),
+    _tensorToVectorReal(0), _configured(false) {
+
+  declareInput(_signal, 4096, "signal", "the input audio signal sampled at 11025 Hz");
+  declareOutput(_predictions, 0, "predictions", "the model predictions");
```
Comment (dbogdanov) [inline src/algorithms/machinelearning/tensorflowpredicttempocnn.cpp:38]: What are these prediction? A more explicit description is better (e.g., `the model predictions (tempo bins activations)`)

File: src/algorithms/spectral/tensorflowinputtempocnn.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+#include "tensorflowinputtempocnn.h"
+
+using namespace std;
+using namespace essentia;
+using namespace standard;
+
+const char* TensorflowInputTempoCNN::name = "TensorflowInputTempoCNN";
+const char* TensorflowInputTempoCNN::category = "Spectral";
+const char* TensorflowInputTempoCNN::description = DOC(
+  "This algorithm computes mel-bands with a particular parametrization specific to Hendrik Schreiber's TempoCNN based models.\n"
```
Comment (dbogdanov) [inline src/algorithms/spectral/tensorflowinputtempocnn.cpp:29]: Same here

Comment (dbogdanov) [reply src/algorithms/spectral/tensorflowinputtempocnn.cpp:29]: Simpler description: "This algorithm computes mel-bands input specific to TempoCNN-based models."  

or 

"This algorithm computes mel-bands specific to the input of TempoCNN-based models."

Is it still clear enough or should we still mention "parametrization "somewhere? (for those wondering why not to use MelBands instead). 

Similar simplification can be done other `TensorflowInput*` algorithms.

File: src/algorithms/rhythm/tempocnn.h [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  ~TempoCNN() {
+    if (_tensorflowPredictTempoCNN) delete _tensorflowPredictTempoCNN;
+  }
+
+  void declareParameters() {
+    declareParameter("graphFilename", "the name of the file containing the model to use", "", Parameter::STRING);
+    declareParameter("input", "the name of the input node in the TensorFlow graph", "", "input");
+    declareParameter("output", "the name of the node from which to retrieve the output tensors", "", "output");
```
Comment (dbogdanov) [inline src/algorithms/rhythm/tempocnn.h:60]: Make sure if the behavior is correct if a user specifies different `output`. As I understand, the intention here is to leave the input / output nodes flexible and up to the user's responsibility.  Should we leave it like that, add a line in the DOC with a note, or have an E_WARNING to make sure the user knows what they do?

Comment (pabloEntropia) [reply src/algorithms/rhythm/tempocnn.h:60]: In the more general `TensorflowPredict`* algorithms `output` is useful to:
1. Match the graph's output node name
2. Retrieve any other node from the graph

While  2) doesn't make sense in `TempoCNN`, 1) is still a valid reason.

For example, different `TempoCNN` models have different output node names in the original implementation (e.g., `output_46`, `output_47`...). Despite we standardize the input and output node names when we serialize the models, `output` would still be useful for a user that serializes the models by himself/herself without renaming the nodes following our convention.

Comment (pabloEntropia) [reply src/algorithms/rhythm/tempocnn.h:60]: I'll rename the parameter →  "the name of the node from which to retrieve the tempo bins activations"

File: src/algorithms/machinelearning/tensorflowpredicttempocnn.cpp (hunk 4 above)
Comment (pabloEntropia) [reply src/algorithms/machinelearning/tensorflowpredicttempocnn.cpp:38]: `TensorflowPredict`* algorithms can be used as feature extractor also depending on the node we retrieve using the `output` parameter.

What about `the output values from the model node named after 'output'`?

File: src/algorithms/rhythm/tempocnn.h (hunk 6 above)
Comment (dbogdanov) [reply src/algorithms/rhythm/tempocnn.h]: Yes, this sounds more clear 👍

File: src/algorithms/machinelearning/tensorflowpredicttempocnn.h [hunk 7]
```diff
@@ -109,7 +109,7 @@ class TensorflowPredictTempoCNN : public Algorithm {
     declareParameter("input", "the name of the input nodes in the Tensorflow graph", "", "input");
     declareParameter("output", "the name of the node from which to retrieve the output tensors", "", "output");
     declareParameter("patchHopSize", "number of frames between the beginnings of adjacent patches. 0 to avoid overlap", "[0,inf)", 128);
-    declareParameter("lastPatchMode", "what to do with the last frames. Options are to `repeat` them to fill the last patch or to discard them", "{discard,repeat}", "discard");
+    declareParameter("lastPatchMode", "what to do with the last frames. `repeat` them to fill the last patch or `discard` them", "{discard,repeat}", "discard");
```
Comment (dbogdanov) [inline src/algorithms/machinelearning/tensorflowpredicttempocnn.h:112]: Should we use `:` consistently? I see some changed lines go for a `.`, others for `:`.

**Review by dbogdanov — APPROVED** [review]
Looks fine to me, just one minor comment for the doc style.

**Pull request merged**